File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
from collections import deque
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        q = deque()
        q.append((entrance[1], entrance[0], 0))
        dx = [-1, 0, 1, 0]
        dy = [0, -1, 0, 1]
        visited = set()
        
        visited.add((entrance[1], entrance[0]))
        
        while q:
            x, y, cnt = q.popleft()

            if cnt != 0 and (x == 0 or y == 0 or x == len(maze[0]) - 1 or y == len(maze) - 1):
                return cnt
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                
                if not (0 <= nx < len(maze[0]) and 0 <= ny < len(maze)):
                    continue
                if maze[ny][nx] == '.' and (nx, ny) not in visited:
                    visited.add((nx, ny))
                    q.append((nx, ny, cnt + 1))
        return -1
                    
        
        

File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.rs
use std::collections::{HashSet, VecDeque};

impl Solution {
    pub fn nearest_exit(maze: Vec<Vec<char>>, entrance: Vec<i32>) -> i32 {
        let mut maze = maze;
        let mut q = VecDeque::new();
        q.push_back((entrance[1] as usize, entrance[0] as usize, 0 as usize));
        maze[entrance[0] as usize][entrance[1] as usize] = '@';
        
        let dx: [isize; 4] = [-1, 0, 1, 0];
        let dy: [isize; 4] = [0, -1, 0, 1];
        
        while !q.is_empty() {
            while let Some((x, y, cnt)) = q.pop_front() {
                if cnt != 0 && (x == 0 || y == 0 || y == maze.len() - 1 || x == maze[0].len() - 1) {
                    return cnt as i32;
                }
                for i in 0..4 {
                    let nx = (x as isize + dx[i]) as usize;
                    let ny = (y as isize + dy[i]) as usize;

                    if nx >= maze[0].len() || nx < 0 || ny >= maze.len() || ny < 0 {
                        continue;
                    }

                    if maze[ny][nx] == '+' || maze[ny][nx] == '@' {
                        continue;
                    }
                    maze[ny][nx] ='@';
                    q.push_back((nx, ny, cnt + 1));
                }
           
            }
        }
        -1
    }
}